src/mixins/chrome/chromeProvider/chromeProvider.ts

```diff
@@ -0,0 +1,54 @@
+import { html, HTMLTemplateResult, nothing, PropertyValues } from "lit";
+import { Component } from "../../mixins";
+import { ChromeAdvertisement, chromeAdvertisementEventName } from "../chromeHost/chromeHost";
+import { state } from "lit/decorators.js";
+import { AbstractComponent } from "../../abstractComponent";
+
+export type ChromeTemplate = HTMLTemplateResult | typeof nothing;
+
+export interface WithChromeProvider {
+  chromeTop?(): ChromeTemplate;
+  chromeBottom?(): ChromeTemplate;
+  chromeLeft?(): ChromeTemplate;
+  chromeRight?(): ChromeTemplate;
+  chromeOverlay?(): ChromeTemplate;
+}
+
+export type ChromeProviderKey = keyof WithChromeProvider;
+
+export const ChromeProvider = <T extends Component>(superClass: T) => {
+  abstract class ChromeProviderComponentClass extends superClass implements WithChromeProvider {
+    @state()
+    private chromeAdvertisement?: ChromeAdvertisement;
+
+    public abstract chromeTop?(): ChromeTemplate;
+    public abstract chromeBottom?(): ChromeTemplate;
+    public abstract chromeLeft?(): ChromeTemplate;
+    public abstract chromeRight?(): ChromeTemplate;
+    public abstract chromeOverlay?(): ChromeTemplate;
```

Comment:
can these be overridden as opposed to abstract and optional?

Reply:
Yeah, I was hoping that TypeScript would be able to type check the return values and provide errors if you didn't implement any of the chrome methods inside an ChromeProvider

As I discussed before, TypeScript doesn't currently support this.

(I'll have to double check), but from memory you can change the return type of overridden functions, meaning that we'd lose type checking here